components/script/script_thread.rs

```diff
@@ -1129,6 +1307,56 @@ impl ScriptThread {
         }
     }
 
+    fn message_to_pipeline(&self, msg: &MixedMessage) -> Option<PipelineId> {
+        match *msg {
+            MixedMessage::FromConstellation(ref inner_msg) => inner_msg.pipeline(),
+            MixedMessage::FromDevtools(ref inner_msg) => inner_msg.pipeline(),
+            MixedMessage::FromScript(_) => None,
+            MixedMessage::FromImageCache((pipeline_id, _)) => Some(pipeline_id),
+            MixedMessage::FromScheduler(ref timer_event) => {
+                let TimerEvent(source, _) = *timer_event;
+                match source {
+                    TimerSource::FromWindow(pipeline_id) => Some(pipeline_id),
+                    _ => None
+                }
+            }
+        }
+    }
+
+    /// Go through documents and see if any don't have a time to interactive yet
+    /// If not, get all tasks for the pipeline that have been running for > 50ms
+    ///    if there are none, then the document is considered interactive
+    ///    otherwise, the window will need to be reset
+    /// Return only after going through all current documents
+    fn check_tti(&self) -> bool {
+        use metrics::MAX_TASK_TIME;
+
+        let times = self.task_times.borrow();
+        let mut all_interactive = true;
+
+        let documents = self.documents.borrow();
+
+        for (pipeline, doc) in documents.iter() {
+            if !doc.is_interactive() {
+                let long_tasks: Vec<&Task> = times.iter()
```

Comment:
I propose a different means of tracking this:
* Each time we start running a task (not just storing it in `sequential`), determine its associated pipeline
* record the start time for the task
* after the task is complete, record the end time for the task
* if the duration is >10ms, record the current time as the start of the document’s TTI window
* for each document that has not had a TTI metric recorded, if current time - TTI window > 50ms then record a TTI metric using the current time
Does that make sense? We should be able to store that information directly in a field on Document, rather than in ScriptThread.